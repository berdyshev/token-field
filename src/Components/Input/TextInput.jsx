import React from 'react';
import PropTypes from 'prop-types';
import { Input } from '@deskpro/react-components';
import styles from '../../styles/style.css';
import TokenInput from './TokenInput';


export default class TextInput extends React.Component {
  static propTypes = {
    token: PropTypes.shape({
      type:  PropTypes.string,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    }).isRequired,
    className:           PropTypes.string,
    onChange:            PropTypes.func,
    selectPreviousToken: PropTypes.func.isRequired,
    selectNextToken:     PropTypes.func.isRequired,
    removeToken:         PropTypes.func.isRequired,
  };
  static defaultProps = {
    className: '',
    onChange() {},
  };

  constructor(props) {
    super(props);
    this.state = {
      value: props.token.value,
    };
  }

  getInput = () => {
    const { value } = this.state;
    return (
      <Input
        value={value}
        className={styles.input}
        onChange={this.handleChange}
        onKeyDown={this.handleKeyDown}
        ref={(c) => { this.input = c; }}
        onFocus={this.moveCaretAtEnd}
        placeholder="________"
      />
    );
  };

  getValue = () => {
    if (this.state.value) {
      return this.state.value;
    }
    return '________';
  };

  focus = () => {
    this.tokenInput.focus();
  };

  focusInput = () => {
    this.input.focus();
  };

  handleChange = (value) => {
    this.setState({
      value
    });
  };

  handleBlur = () => {
    this.props.onChange(this.state.value);
  };

  handleKeyDown = (e) => {
    switch (e.key) {
      case 'Escape':
        this.setState({
          value: this.props.token.value
        }, () => {
          this.tokenInput.disableEditMode();
        });
        break;
      case 'Tab':
        if (e.shiftKey) {
          this.props.selectPreviousToken();
        } else {
          this.props.selectNextToken();
        }
        this.tokenInput.disableEditMode();
        break;
      case 'Enter':
        this.props.selectNextToken();
        this.tokenInput.disableEditMode();
        break;
      default:
        return true;
    }
    e.preventDefault();
    return true;
  };

  moveCaretAtEnd = () => {
    const input = this.input.input;
    if (this.state.value) {
      const length = this.state.value.length;
      input.setSelectionRange(length, length);
    }
  };

  render() {
    const { token, className, removeToken } = this.props;
    return (
      <TokenInput
        ref={(c) => { this.tokenInput = c; }}
        className={className}
        type={token.type}
        renderInput={this.getInput}
        renderValue={this.getValue}
        onFocus={this.focusInput}
        onBlur={this.handleBlur}
        removeToken={removeToken}
      />
    );
  }
}
