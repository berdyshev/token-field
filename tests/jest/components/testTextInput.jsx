import React from 'react';
import { shallow } from 'enzyme';
import ReactDOM from 'react-dom';
import TextInput from 'Components/Input/TextInput';

let wrapper;
let token;

beforeEach(() => {
  token = {
    type:  'user-message',
    value: 'help upgrading'
  };
  wrapper = shallow(<TextInput token={token} className="test" />);
});

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<TextInput token={token} />, div);
});

it('should render the input label', () => {
  expect(wrapper.contains(token.type)).toEqual(true);
});

it('should render the input value', () => {
  expect(wrapper.contains(token.value)).toEqual(true);
});

it('should have the class passed', () => {
  expect(wrapper.first().hasClass('test')).toEqual(true);
});
