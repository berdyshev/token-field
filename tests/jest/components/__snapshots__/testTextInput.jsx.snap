// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`+++capturing Snapshot of TextInput 1`] = `
<div
  className="token test"
>
  <div
    className="dp-code label"
  >
    user-message
    :
  </div>
  <span
    className="value"
    onClick={[Function]}
  >
    help upgrading
  </span>
  <div
    className="token-remove remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;

exports[`+++capturing Snapshot of TextInput with empty value 1`] = `
<div
  className="token test"
>
  <div
    className="dp-code label"
  >
    user-message
    :
  </div>
  <span
    className="value"
    onClick={[Function]}
  >
    ________
  </span>
  <div
    className="token-remove remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;
